firewood/src/storage/linear/proposed.rs

```diff
@@ -0,0 +1,526 @@
+    Initial,
+    // we know we are not in a modified area, so find the next modified area
+    FindNext,
+    // we know there are no more modified areas past our current offset
+    NoMoreModifiedAreas,
+    BeforeModifiedArea {
+        next_offset: u64,
+        next_modified_area: &'a [u8],
+    },
+    InsideModifiedArea {
+        area: &'a [u8],
+        offset_within: usize,
+    },
+}
+
+impl<P: ReadLinearStore, M: Send + Sync + Debug> ReadLinearStore for Proposed<P, M> {
+    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
+        Ok(Box::new(LayeredReader {
+            offset: addr,
+            state: LayeredReaderState::Initial,
+            layer: self,
+        }))
+    }
+
+    fn size(&self) -> Result<u64, Error> {
+        // start with the parent size
+        let parent_size = self.parent.state.size()?;
+        // look at the last delta, if any, and see if it will extend the file
+        Ok(self
+            .new
+            .range(..)
+            .next_back()
+            .map(|(k, v)| *k + v.len() as u64)
+            .map_or_else(
+                || parent_size,
+                |delta_end| std::cmp::max(parent_size, delta_end),
+            ))
+    }
+}
+
+/// Marker that the Proposal is mutable
+#[derive(Debug)]
+pub(crate) struct Mutable;
+
+/// Marker that the Proposal is immutable
+#[derive(Debug)]
+pub(crate) struct Immutable;
+
+impl<P: ReadLinearStore> WriteLinearStore for Proposed<P, Mutable> {
+    // TODO: we might be able to optimize the case where we keep adding data to
+    // the end of the file by not coalescing left. We'd have to change the assumption
+    // in the reader that when you reach the end of a modified region, you're always
+    // in an unmodified region
+
+    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error> {
+        // the structure of what will eventually be inserted
+        struct InsertData {
+            offset: u64,
+            data: Box<[u8]>,
+        }
+
+        debug_assert!(!object.is_empty());
+        let mut insert = InsertData {
+            offset,
+            data: Box::from(object),
+        };
+
+        // we track areas to delete here, which happens when we combine
+        let mut merged_offsets_to_delete = Vec::new();
+
+        // TODO: combining a change where we constantly append to a file might be too
+        // expensive, so consider skipping the relocations if they don't overlap at all
+        for existing in self.new.range(..=insert.offset).rev() {
+            let old_end = *existing.0 + existing.1.len() as u64;
+            // found    smme--------- s=existing.0 e=old_end
+            // original ----smmme---- s=new_start e=new_start+object.len() [0 overlap]
+            // original --smmmme----- with overlap
+            // new      smmmmmmme----
+            if insert.offset <= old_end {
```

Comment:
Should this be `<` instead of `<=`?

IIUC `old_end` is the first index _after_ the `existing` diff ends. For example, if `existing.0` and `existing.1` are `0` and `&[1,2,3]` respectively, then `old_end` is `3`, and index `3` is after this diff. And IIUC `insert.offset` is the first index affected by the diff. So if `insert.offset == old_end` then these diffs don't overlap.